crates/fuel-core/src/service/adapters/graphql_api.rs

```diff
@@ -223,6 +236,40 @@ impl TxPoolPort for TxPoolAdapter {
     }
 }
 
+impl DatabaseMessageProof for Database {
+    fn block_history_proof(
+        &self,
+        message_block_height: &BlockHeight,
+        commit_block_height: &BlockHeight,
+    ) -> StorageResult<MerkleProof> {
+        if message_block_height > commit_block_height {
```

Comment:
This seems like a lot of business logic that is being wrapped inside of an adapter. Ideally, this logic would exist in a more unit testable domain context such as one of the following:


- A query context for merkle proofs that just fulfills this particular API. The ports would need the lower-level functions like fetching merkle data that do minimal logic.
- Implemented on the fuel-core database itself

Reply:
> A query context for merkle proofs that just fulfills this particular API. The ports would need the lower-level functions like fetching merkle data that do minimal logic.

It is not actually a lot of business logic because it only takes data from the table and passes it into `MerkleTree`. I put it here because I don't want to expose the `FuelBlockMerkleMetadata` and `FuelBlockMerkleData` tables outside the `fuel-core`(The `MerkleTree` requires explicit work with `FuelBlockMerkleData`). 

Also, it aligns with the current code for blocks. It is an internal detail of how the merklezation of the database works and we can change it at any time.

> Implemented on the fuel-core database itself

I will move it to `Database` itself=)